vital/types/category_hierarchy.cxx

```diff
@@ -0,0 +1,365 @@
+category_hierarchy::label_id_t
+category_hierarchy
+::get_class_id( const label_t& class_name ) const
+{
+  hierarchy_const_itr_t itr = this->find( class_name );
+
+  return itr->second->category_id;
+}
+
+
+// -----------------------------------------------------------------------------
+category_hierarchy::label_vec_t
+category_hierarchy
+::get_class_parents( const label_t& class_name ) const
+{
+  label_vec_t output;
+
+  hierarchy_const_itr_t itr = this->find( class_name );
+
+  for( category *p : itr->second->parents )
+  {
+    output.push_back( p->category_name );
+  }
+
+  return output;
+}
+
+
+// -----------------------------------------------------------------------------
+void
+category_hierarchy
+::add_relationship( const label_t& child_name, const label_t& parent_name )
+{
+  hierarchy_const_itr_t itr1 = this->find( child_name );
+  hierarchy_const_itr_t itr2 = this->find( parent_name );
+
+  itr1->second->parents.push_back( itr2->second.get() );
+  itr2->second->children.push_back( itr1->second.get() );
+}
+
+
+// -----------------------------------------------------------------------------
+void
+category_hierarchy
+::add_synonym( const label_t& class_name, const label_t& synonym_name )
+{
+  hierarchy_const_itr_t itr = this->find( class_name );
+
+  if( has_class_name( synonym_name ) )
+  {
+    throw std::runtime_error( "Synonym name already exists in hierarchy" );
+  }
+
+  itr->second->synonyms.push_back( synonym_name );
+  m_hierarchy[ synonym_name ] = itr->second;
+}
+
+
+// -----------------------------------------------------------------------------
+category_hierarchy::label_vec_t
+category_hierarchy
+::all_class_names() const
+{
+  std::vector< category_sptr > sorted_cats = sorted_categories();
+
+  label_vec_t names;
+
+  for( category_sptr c : sorted_cats )
+  {
+    names.push_back( c->category_name );
+  }
+
+  return names;
+}
+
+
+// -----------------------------------------------------------------------------
+category_hierarchy::label_vec_t
+category_hierarchy
+::child_class_names() const
+{
+  std::vector< category_sptr > sorted_cats = sorted_categories();
+
+  label_vec_t names;
+
+  for( category_sptr c : sorted_cats )
+  {
+    if( c->children.empty() )
+    {
+      names.push_back( c->category_name );
+    }
+  }
+
+  return names;
+}
+
+
+// -----------------------------------------------------------------------------
+size_t
+category_hierarchy
+::size() const
+{
+  return m_hierarchy.size();
+}
+
+
+// -----------------------------------------------------------------------------
+void
+category_hierarchy
+::load_from_file( std::string filename )
+{
+  std::ifstream in( filename.c_str() );
+
+  if( !in )
+  {
+    throw std::runtime_error( "Unable to open " + filename );
+  }
+
+  std::vector< std::pair< label_t, label_t > > relationships;
+
+  std::string line;
+  label_t label;
+
+  int entry_num = 0;
+
+  while( std::getline( in, line ) )
```

Comment:
You can use the vital/util/data_stream_reader here.

    kwiver::vital::data_stream_reader dsr( in );
    while( dsr.getline( line ) )

The data stream reader removes shell comments and blank lines.